File: controller/health.controller.js
/**
 * System controller.
 *
 * Handlers for the unversioned system endpoints: backend info (GET /)
 * and the health check (GET /health).
 */
const asyncHandler = require("../utils/asyncHandler");
const db = require('../config/db');
const config = require('../config/config');
const { Env } = require('../utils/constant');
const { name, version, description } = require('../package.json');

/**
 * GET /
 * Returns basic backend info (name/version/description from package.json,
 * current environment, and useful links). The Swagger docs link is only
 * included outside production, where the /api-docs route actually exists.
 */
exports.getInfo = asyncHandler(async (req, res) => {
    const info = {
        name,
        version,
        description,
        environment: config.env,
        apiBaseUrl: '/api/v1',
        health: '/health',
    };

    if (config.env !== Env.production) {
        info.docs = '/api-docs';
    }

    return res.send(info);
})

/**
 * GET /health
 * Health check for uptime monitors / load balancers. Pings the database
 * with `db.authenticate()` and responds:
 * - 200 { status: 'ok' }       → server and database both reachable
 * - 503 { status: 'degraded' } → server up but database unreachable
 */
exports.getHealth = asyncHandler(async (req, res) => {
    let database = 'up';

    try {
        await db.authenticate();
    } catch (error) {
        database = 'down';
    }

    const healthy = database === 'up';

    return res.status(healthy ? 200 : 503).send({
        status: healthy ? 'ok' : 'degraded',
        server: 'up',
        database,
        uptime: process.uptime(),
        timestamp: new Date().toISOString(),
    });
})

File: controllers/health_controller.py
"""
System controller.

Handlers for the unversioned system endpoints: backend info (GET /)
and the health check (GET /health). Registered as view functions from
routes/health_routes.py — Flask dispatches any raised exception straight to
the global error handler (middlewares/error.py), so no asyncHandler-style
wrapper is needed here (unlike the original Express version).
"""
from datetime import datetime, timezone

from flask import jsonify
from sqlalchemy import text

from config.config import config, APP_NAME, APP_VERSION, APP_DESCRIPTION
from config.db import db
from utils.constant import Env

# process start time, used to compute uptime for the health check
_start_time = datetime.now(timezone.utc)


def get_info():
    info = {
        'name': APP_NAME,
        'version': APP_VERSION,
        'description': APP_DESCRIPTION,
        'environment': config.ENV,
        'apiBaseUrl': '/api/v1',
        'health': '/health',
    }

    if config.ENV != Env.PRODUCTION:
        info['docs'] = '/api-docs'

    return jsonify(info)


def get_health():
    database = 'up'

    try:
        db.session.execute(text('SELECT 1'))
    except Exception:
        database = 'down'
        # a failed connection/query can leave the scoped session's
        # transaction in a broken state; without this, Flask-SQLAlchemy's
        # end-of-request teardown (db.session.remove()) can itself raise
        # trying to clean up that broken state — an error that happens
        # *after* this view already returned, outside any @app.errorhandler
        db.session.rollback()

    healthy = database == 'up'
    uptime_seconds = (datetime.now(timezone.utc) - _start_time).total_seconds()

    return jsonify({
        'status': 'ok' if healthy else 'degraded',
        'server': 'up',
        'database': database,
        'uptime': uptime_seconds,
        'timestamp': datetime.now(timezone.utc).isoformat(),
    }), (200 if healthy else 503)
